codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java

```diff
@@ -416,6 +416,9 @@ private static HeaderType validate(int streamId, CharSequence name,
             throw streamError(streamId, PROTOCOL_ERROR,
                     "Illegal value specified for the 'TE' header (only 'trailers' is allowed).");
         }
+        if (!isValidHeaderValue(value)) {
```

Comment:
I'm agreed that this validation is important and we should follow the spec. On the other hand, header values are tend to be quite long. Validating each value may be a significant performance hit. There are cases when communication is trusted between peers and such validation is not necessary. Can we consider an opt-out config options from this validation?

Reply:
We have a config for validating the headers. You want a separate one for header _value_ validation?

Reply:
Yes, validation of the header-name is more important and is used by many netty users by default (header names usually has more predictable length). If we add more validation for header values, we should let users preserve pre-existing behavior of validating only header names but not values.

Reply:
Actually, since the same validation rule applies for HTTP/2 and HTTP/1.x, have you considered moving this validation to the upper layer where it can be shared between two?

Reply:
ServiceTalk already does validation of the header values for both protocols, if netty also does it internally for HTTP/2 only without option to opt-out or share validation between HTTP/2 and HTTP/1.x, ServiceTalk users may end up doing expensive validation 2 times.

Reply:
I would leave it up to @normanmaurer. The benefit of doing this now instead of a separate PR is the API impact. For example, if you end up moving values validation to `DefaultHeaders` to share that between HTTP/2 and HTTP/1.X, you won't need any API changes for `DefaultHttp2HeadersDecoder` made in this PR.

Btw, moving this validation to `DefaultHeaders` has another benefit: users will be able to control it for inbound and outbound messages. In my practice, users frequently make mistakes in outbound header values, like adding `\n` at the end of the value (especially when the value is pulled from somewhere). When an illegal header sent, they see a weird error at transport level that doesn't describe the cause for request failure. Very hard to debug unless you have a validation before sending the request that describes where is the mistake.